test/helpers/checkUpgrades.ts

```diff
@@ -0,0 +1,64 @@
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { ZNSRegistry, ZNSRegistryMock  } from "../../typechain";
+import { BigNumber, ethers } from "ethers";
+
+export const registryActions = async (
+  registry : ZNSRegistry,
+  domainHash : string,
+  accounts : Array<SignerWithAddress>,
+) => {
+  const [deployer, operator, mockRegistrar, mockResolver] = [...accounts];
+
+  // Add an operator
+  await registry.connect(deployer).setOwnerOperator(operator.address, true);
+
+  // Create a domain record
+  await registry.connect(mockRegistrar).createDomainRecord(
+    domainHash,
+    deployer.address,
+    mockResolver.address
+  );
+};
+
+export const registryPromises = async (
+  registry : ZNSRegistry,
+  domainHash : string,
+  accounts : Array<SignerWithAddress>
+) => {
+  const [deployer, operator] = [...accounts];
+
+  const calls = [
+    registry.isOwnerOrOperator(domainHash, deployer.address),
+    registry.isOwnerOrOperator(domainHash, operator.address),
+    registry.exists(domainHash),
+    registry.getAccessController(),
+  ];
+
+  return Promise.all(calls);
+};
+
+export const upgradedRegistryActions = async (
+  upgradedRegistry : ZNSRegistryMock,
+  newNumber : BigNumber,
+  randomUser : SignerWithAddress
+) => {
```

Comment:
maybe we can change the helpers to a different style:
we can make these more abstract to work with any contract, and this will check values from arg1 to the values from arg2, both of which can be contract specific callbacks that we pass as an array for each arg.
we specify the calls as args in the test itself, this way it will be easily readable of what exactly we are checking in the test, while the helper will make these calls and do comparisons internally. we don't need to see comparisons in the tests as much as what we are checking exactly, which we can tell by the callbacks we pass to the helper in the test body.
what do you think?